### PR TITLE
Update implementation of all checked and saturating arithmetic operations

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

*Description of changes:*

Check for potential overflows before they occur. (https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/common/array_list.inl [hunk 1]
```diff
@@ -258,7 +258,7 @@ int aws_array_list_set_at(struct aws_array_list *AWS_RESTRICT list, const void *
     /* this isn't perfect but its the best I can come up with for detecting
      * length changes*/
     if (index >= aws_array_list_length(list)) {
-        list->length = index + 1;
+        aws_add_size_checked(index, 1, &list->length);
```
Comment (feliperodri) [inline include/aws/common/array_list.inl]: Our proof also catch this overflow.

Comment (danielsn) [reply include/aws/common/array_list.inl]: This should be guarded by a check.

File: .cbmc-batch/jobs/aws_mul_size_checked/aws_mul_size_checked_harness.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    return AWS_OP_SUCCESS;
+}
+
+void aws_mul_size_checked_harness() {
+    size_t r = nondet_size_t();
+    size_t __r = nondet_size_t();
+    __CPROVER_assume(r == __r);
+
+    if(nondet_int()){
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_mul_size_checked/aws_mul_size_checked_harness.c]: Maybe add a comment that doing it with the full range of undet inputs leads to excessively long runtimes?

File: include/aws/common/math.fallback.inl [hunk 3]
```diff
@@ -1,116 +1,100 @@
...
-    return x;
+    return a + b;
 }
 
 /**
- * Adds a + b and returns the result in *r. If the result
- * overflows, returns AWS_OP_ERR; otherwise returns AWS_OP_SUCCESS.
- */
+* If a + b overflows, returns AWS_OP_ERR; otherwise adds
+* a + b, returns the result in *r, and returns AWS_OP_SUCCESS.
```
Comment (danielsn) [inline include/aws/common/math.fallback.inl]: The signatures on the other math.* files should be updated to match

**Pull request closed without merging**